[WIP][Not for review] Added skeleton of python dispatch key

Chillee opened the pull request:
Successfully threads a PyObject* through the dispatcher as an attribute of `PythonTensorImpl` and calls back into its `__torch_function__`.

Starts to implement this: https://github.com/pytorch/pytorch/issues/55095

Unresolved issues:
1. How do I map from `aten::<op>` back to a python op, and pass the arguments as well?
2. I need to add some kind of shallow copy method for autograd.

Chillee committed:
Added skeleton of python dispatch key

zou3519 commented:
> 1. How do I get the torch_api_function from the c10::OperatorHandle?

One way to hack around it could be to add a new torch._ops namespace that contains python bindings for the entirety of the operator library. But there is a larger problem that FX doesn't want to see internal pytorch operators.

Chillee committed:
made some more progress on boxing/unboxing

Chillee committed:
successfully traced through autograd!!!

Chillee committed:
modified thing

Chillee committed:
stashing progress

Chillee commented:
An example of tracing through grad
![image](https://user-images.githubusercontent.com/6355099/115177512-63d2c880-a09d-11eb-9946-3683cfcb9bc6.png)

and tracing through vmap

![image](https://user-images.githubusercontent.com/6355099/115179483-66372180-a0a1-11eb-8fac-88333952dda7.png)

Chillee committed:
Added FX tracing through autograd example

Chillee commented:
Added an example of FX tracing through autograd (and then JIT tracing through FX :P).

![image](https://user-images.githubusercontent.com/6355099/115544534-b6b0a980-a270-11eb-8852-aaccec9f32c0.png)

Chillee committed:
added some benchmarking

Chillee committed:
moved computeNOperand operations over

ghstack-source-id: c4d0e7c76fd5004eef2756ca18c1e2e6419ecb69
Pull Request resolved: https://github.com/pytorch/pytorch/pull/55372

moved some more stuff over

did more refactoring

Chillee committed:
Added matmul/unified dtypes

ghstack-source-id: 781d73f88440e49d02be63a82d542c955444afb4
Pull Request resolved: https://github.com/pytorch/pytorch/pull/56456

Chillee committed:
moved lowerings out of the TensorExprKernel and into independent functions

ghstack-source-id: 1778647df624f30b82f1bc02aedb9ea2fcf3f483
Pull Request resolved: https://github.com/pytorch/pytorch/pull/56679

moved lowerings out of the TensorExprKernel and into independent functions

ghstack-source-id: 1778647df624f30b82f1bc02aedb9ea2fcf3f483
Pull Request resolved: https://github.com/pytorch/pytorch/pull/56680

Chillee committed:
Added full gradient compilation example

Chillee committed:
stuff
